cfgov/login/auth.py

```diff
@@ -0,0 +1,158 @@
+import logging
+
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import Group, Permission
+from django.contrib.contenttypes.models import ContentType
+from django.utils.encoding import smart_str
+from django.utils.module_loading import import_string
+
+from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+
+
+User = get_user_model()
+
+logger = logging.getLogger(__name__)
+
+
+def username_from_email(email, claims=None):
+    """Generate a username for new users from their email address
+
+    If this function is called, it's already tried to look up an email
+    address to match a user and failed. But if the user portion of the email
+    address already exists as a user, this function will try to find an
+    unused username by appending a counter to the end."""
+    email_username = smart_str(email.split("@")[0])
+
+    username = email_username
+    counter = 1
+    while User.objects.filter(username=username).exists():
+        username = email_username + str(counter)
+        counter += 1
+
+    return username
+
+
+def process_roles_admin(user, roles):
+    """Adjust a user's is_superuser property based on OIDC roles claim"""
+    user_modified = False
+
+    if settings.OIDC_OP_ADMIN_ROLE is None:
+        logger.debug("OIDC_OP_ADMIN_ROLE not set, not modifying is_superuser")
+        return
+
+    is_admin = settings.OIDC_OP_ADMIN_ROLE in roles
+
+    if is_admin and not user.is_superuser:
+        logger.debug(f"Setting is_superuser for {user.username}.")
+        user.is_superuser = True
+        user_modified = True
+    elif not is_admin and user.is_superuser:
+        logger.debug(f"Removing is_superuser from {user.username}.")
+        user.is_superuser = False
+        user_modified = True
+
+    return user_modified
+
+
+def process_roles_user(user, roles):
+    """Adjust user group assignment based on OIDC roles"""
+    user_modified = False
+
+    if settings.OIDC_OP_USER_ROLE is None:
+        logger.debug("OIDC_OP_USER_ROLE not set, not modifying user groups")
+        return
+
+    if settings.SSO_DEFAULT_GROUP is None:
+        logger.debug("SSO_DEFAULT_GROUP not set, not modifying user groups")
+        return
+
+    # All SSO users will belong to a group that can log into the admin
+    group = Group.objects.get_or_create(name=settings.SSO_DEFAULT_GROUP)[0]
+
+    # Make sure this group has permission to log into the admin
+    permission = Permission.objects.get(
+        codename="access_admin",
+        content_type=ContentType.objects.get(
+            app_label="wagtailadmin", model="admin"
+        ),
+    )
+    if not group.permissions.contains(permission):
+        group.permissions.add(permission)
```

Comment:
Could this logic maybe be moved to a data migration, to avoid the need to run it each time? Wagtail built-in groups and permissions tend to be managed via migrations.

Reply:
That's a good idea.